### PR TITLE
add telemetry reporting module

### PR DESCRIPTION
**Pull request by kejxu**
Brian (from IoT team) is also driving to explore if it's possible for us to send telemetry to an internal centralized App Insights pool. We could (easily) switch to that route by changing the `aiKey` defined in `package.json` once we understand how that could be done.

### CODE REVIEW

File: src/vscode-utils.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+export interface IPackageInfo {
+    name: string;
+    version: string;
+    aiKey: string;
+}
+
+export function getPackageInfo(context: vscode.ExtensionContext): IPackageInfo {
+    const metadataFile: string = "package.json";
+    const extensionMetadata = require(path.join(context.extensionPath, metadataFile));
```
Comment (ooeygui) [inline src/vscode-utils.ts]: Let's gracefully fail instead of crashing the extension.

File: src/extension.ts [hunk 2]
```diff
@@ -56,14 +57,18 @@ export enum Commands {
 }
 
 export async function activate(context: vscode.ExtensionContext) {
+    const logger: telemetry.ILogger = telemetry.getLogger(context);
+
     // Activate if we're in a catkin workspace.
     await determineBuildSystem(vscode.workspace.rootPath);
 
     if (buildSystem == BuildSystem.None) {
+        logger.logActivate("skipped");
```
Comment (ooeygui) [inline src/extension.ts]: We shouldn't log skips - nothing we can do with that data.

File: src/extension.ts [hunk 3]
```diff
@@ -56,14 +57,18 @@ export enum Commands {
...
         return;
     }
 
     console.log(`Activating ROS extension in "${baseDir}"`);
+    logger.logActivate("found_catkin");
```
Comment (ooeygui) [inline src/extension.ts]: Remove this until we have other build system types

File: src/extension.ts [hunk 4]
```diff
@@ -148,16 +155,46 @@ function activateEnvironment(context: vscode.ExtensionContext) {
...
-        vscode.commands.registerCommand(Commands.ShowCoreStatus, () => { master.launchMonitor(context) }),
-        vscode.commands.registerCommand(Commands.StartRosCore, master.startCore),
-        vscode.commands.registerCommand(Commands.TerminateRosCore, () => { master.stopCore(masterApi) }),
-        vscode.commands.registerCommand(Commands.UpdateCppProperties, build.updateCppProperties),
-        vscode.commands.registerCommand(Commands.UpdatePythonPath, build.updatePythonPath),
-        vscode.commands.registerCommand(Commands.Rosrun, rosrundelegate),
-        vscode.commands.registerCommand(Commands.Roslaunch, roslaunchdelegate),
+        vscode.commands.registerCommand(Commands.CreateCatkinPackage, () => {
+            logger.logCommand(Commands.CreateCatkinPackage);
+            catkin.createPackage();
```
Comment (ooeygui) [inline src/extension.ts]: Prefer logging within the function rather than having the caller log on its behalf.

File: package.json [hunk 5]
```diff
@@ -11,6 +11,7 @@
     "displayName": "ROS",
     "description": "Develop Robot Operating System (ROS) with Visual Studio Code.",
     "icon": "media/icon.png",
+    "aiKey": "6a0f99ed-47cf-4ce0-b2e5-321ef7a6aa67",
```
Comment (ooeygui) [inline package.json]: Put this in the pipeline instead of in github.

Comment (ooeygui) [reply package.json]: Use a different key for pipeline builds from github builds.

Comment (kejxu) [reply package.json]: created https://github.com/ms-iot/vscode-ros/issues/60 to track this

File: src/ros/rosrun.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+export async function setup(context: vscode.ExtensionContext) {
+    const reporter = telemetry.getReporter(context);
+    reporter.sendTelemetryCommand(extension.Commands.Rosrun);
+
+    const terminal = await preparerosrun();
+    terminal.show();
+}
+
+async function preparerosrun(): Promise<vscode.Terminal> {
```
Comment (ooeygui) [inline src/ros/rosrun.ts]: Camel Case

**Pull request merged**